src/opensearch.py

```diff
@@ -0,0 +1,195 @@
+        if not self._opensearch.present:
+            raise OpenSearchMissingError()
+
+        try:
+            self._opensearch.stop([self.SERVICE_NAME])
+        except SnapError as e:
+            logger.error(f"Failed to stop the opensearch.{self.SERVICE_NAME} service. \n{e}")
+            raise OpenSearchStopError()
+
+    def restart(self):
+        """Restart the snap exposed "daemon" service."""
+        if not self._opensearch.present:
+            raise OpenSearchMissingError()
+
+        try:
+            self._opensearch.restart([self.SERVICE_NAME])
+        except SnapError as e:
+            logger.error(f"Failed to restart the opensearch.{self.SERVICE_NAME} service. \n{e}")
+            raise OpenSearchRestartError()
+
+    def _build_paths(self) -> Paths:
+        """Builds a Path object.
+
+        The main paths are:
+          - OPENSEARCH_HOME: read-only path ($SNAP/..), where the opensearch binaries are
+          - OPENSEARCH_CONF: writeable by root or snap_daemon ($SNAP_COMMON) where config files are
+        """
+        return Paths(
+            home="/var/snap/opensearch/current",
+            conf="/var/snap/opensearch/common/config",
+            data="/var/snap/opensearch/common/data",
+            logs="/var/snap/opensearch/common/logs",
+            jdk="/var/snap/opensearch/current/jdk",
+            tmp="/var/snap/opensearch/common/tmp",
+        )
+
+
+class OpenSearchTarball(OpenSearchDistribution):
+    """Tarball distro of opensearch, only overrides properties and logic proper to the tar."""
+
+    def __init__(self, charm, peer_relation: str):
+        super().__init__(charm, peer_relation)
+
+    def install(self):
+        """Temporary (will be deleted later) - Download and Un-tar the opensearch distro."""
+        try:
+            response = requests.get(
+                "https://artifacts.opensearch.org/releases/bundle/opensearch/2.3.0/opensearch-2.3.0-linux-x64.tar.gz"
+            )
+
+            tarball_path = "opensearch.tar.gz"
+            with open(tarball_path, "wb") as f:
+                f.write(response.content)
+        except Exception as e:
+            logger.error(e)
+            raise OpenSearchInstallError()
+
+        extract_tarball(tarball_path, self.paths.home)
+
+    def start(self):
+        """Start opensearch as a Daemon."""
+        logger.debug("Starting opensearch.")
+        if self.is_started():
+            return
+
+        try:
+            self._setup_linux_perms()
+            self._run_cmd(
+                "setpriv",
+                f"--clear-groups --reuid ubuntu --regid ubuntu -- {self.paths.home}/bin/opensearch --daemonize",
+            )
+        except OpenSearchCmdError:
+            raise OpenSearchStartError()
+
+        if not self.is_started():
+            raise OpenSearchStartError()
+
+    def stop(self):
+        """Stop opensearch."""
+        self._run_cmd("ps aux | grep opensearch | awk '{print $2}' | kill -15")
```

Comment:
you can just use `pkill` with the same effect
```suggestion
        self._run_cmd("pkill -15 opensearch")
```